initial implementation of rb_common

halorrr opened the pull request:
This is my initial implementation of the rb_common shared interfaces similar to what we have available for python scrapers. I started with a 1:1 of the python design and the refactored from there to what I feel is a better long term design. Both the config and graphql have a base class so that different child classes (for example on configs I now have a separate one for Stash and StashDB, where previously there was just one for stash) can have a common API. This also worked will since `stash_graphql` is now a child class, and the base query method is in the base class so that other common api interface (maybe a stashdb one in the future) can share that method.

The graphql interface has a third party dependancy on the `faraday` gem which I have included in my PR that adds Ruby to stash https://github.com/stashapp/stash/pull/2707 (that should be included in the 0.17 release)

Within the stash graphql interface I've moved any of the written out graphql queries to private methods on the class that are class by public methods like:
```Ruby
    def configuration
      query(configuration_query)["configuration"]
    end
```

Found this made the available methods in the API much easier to read. In a future PR I will work on adding a shared interface for comparing and adding scene markers during a scrape.

I included REAME files at every level so that there is clear documentation on how everything is currently set up and how to use it.

bnkai commented on scrapers/rb_common/logger.rb:
```diff
@@ -0,0 +1,47 @@
+    def log(level, text)
+      level_char = control_wrap(level)
+
+      # I'm not sure what case is covered by the image part of this regex, but it
+      # was present in the py_common version so I've included it.
```
There was an issue when trying to print to the console through the log function a base64 encoded image. It could be python's issue or a console print limitation. Eitherway it caused the scrapers to error out so every attemt to log print a base64 scraped image was replaced with `[...]`

halorrr replied:
Hmmm I'll test that out, cause if it isn't needed, I'd love to remove it, and if it is needed, I'll add a proper comment here with why it is needed.

halorrr replied:
Do you know of a test case where this would happen? Having trouble finding a way to replicate

bnkai commented:
Just for testing purposes can you provide a sample scraper as well ?

halorrr commented:
@bnkai For sure. You will ask need to run this in the container to get ruby and faraday installed `apk add --no-cache ruby && gem install faraday` (which now that I think about, I'll add to the readme until ruby is officially added to stash.

The main one that I have worked on so far is the ruby port / improvement of the VixenNetwork script. This is the yaml for it:
```YAML
name: "Vixen Media Group"
sceneByURL:
  - url:
      - blacked.com/videos
      - blackedraw.com/videos
      - deeper.com/videos
      - slayed.com/videos
      - tushy.com/videos
      - tushyraw.com/videos
      - vixen.com/videos
    action: script
    script:
      - ruby
      - vixenNetwork.rb

# If your search query contains the studio name, the scraper will only search on that site.
# 'Ariana Vixen Cecilia' will only search on Vixen for 'Ariana Cecilia`
# If your search query starts with a '$' all sites will be searched
# '$Ariana Vixen Cecilia' will search in all sites for 'Ariana Vixen Cecilia'
sceneByName:
    action: script
    script:
      - ruby
      - vixenNetwork.rb
      - search
sceneByQueryFragment:
    action: script
    script:
      - ruby
      - vixenNetwork.rb

# Last Updated July 05, 2022
```

And this is the script so far:
```Ruby
# frozen_string_literal: true

require_relative "rb_common/rb_common"

module VixenMediaGroup
  class Config
    SEARCH_OPTIONS = {
      # Max scenes to return per search when using sceneByName
      max_scenes: 6
    }
    MARKER_OPTIONS = {
      # Would you like to add markers to scenes when scraping?
      create_markers: false,
      # Only create marker if the durations match (API vs Stash) or are a least
      # within the acceptable range set in max_duration_difference
      marker_duration_match: true,
      # Maximum second differnce allowed in scene length between Stash & API
      # This setting only applies if marker_duration_match is turned on.
      max_duration_difference: 15
    }
    TAG_OPTIONS = {
      # List of tags you would like to not be be ignored when scraping. Will not ignore any fixed tags set below.
      ignored_tags: [],
      # List of tags you would like to always show up on scrape even when not returned by studio scrape
      fixed_tags: []
    }

    def self.max_scenes
      SEARCH_OPTIONS[:max_scenes]
    end

    def self.create_markers?
      MARKER_OPTIONS[:create_markers]
    end

    def self.marker_duration_match?
      MARKER_OPTIONS[:marker_duration_match]
    end

    def self.max_duration_difference
      MARKER_OPTIONS[:max_duration_difference]
    end

    def self.ignored_tags
      TAG_OPTIONS[:ignored_tags]
    end

    def self.fixed_tags
      format_tags(TAG_OPTIONS[:fixed_tags])
    end

    def self.fixed_tags?
      !TAG_OPTIONS[:fixed_tags].empty?
    end

    private

    def self.format_tags(tags)
      tags.map { |tag| { "name" => tag } }
    end
  end

  class Site
    MAX_SCENES = VixenMediaGroup::Config.max_scenes

    attr_reader :name, :id, :home

    def initialize(name)
      @name = name
      @id = name.gsub(" ", "").upcase
      @api = "https://www." + @id.downcase + ".com/graphql"
      @home = "https://www." + @id.downcase + ".com"
      @search_count = MAX_SCENES
      @logger = Stash::Logger.new
      @config = VixenMediaGroup::Config
    end

    def get_scene(url)
      @logger.debug("Scraping using #{@name} GraphQL API")
      variables = {
        "site": @id,
        "videoSlug": scene_slug(url)
      }

      response = vixen_query(get_video_query, variables)
      parse_scene(response, url)
    end

    def get_search_result(search)
      @logger.debug("Searching using #{@name} GraphQL API")
      variables = {
        "site": @id,
        "query": search,
        "first": @search_count
      }

      response = vixen_query(get_search_query, variables)
      parse_search(response)
    end

    def valid_url?(url)
      url = url.downcase.gsub(/\/$/, "")
      parsed_url = URI.parse(url)

      return false if parsed_url.host.nil?

      if parsed_url.host.gsub(/www.|\.com/, "") == @id.downcase
        parts = url.split("/")
        return false if parts.count < 4
        return true if parts[-2] == "videos"
        false
      else
        false
      end
    end

    def length
      @id.length
    end

    private

    def vixen_query(query, variables = nil)
      headers = {
        "Content-Type": "application/json",
        "Accept": "application/json",
        "DNT": "1",
      }
      connection = Faraday.new(url: @api, headers: headers)
      response = connection.post do |request|
        body = { "query" => query }
        body["variables"] = variables if variables

        request.body = body.to_json
      end

      case response.status
      when 200
        result = JSON.parse(response.body)
        if result["error"]
          result["error"]["errors"].each do |error|
            logger.error("GraphQL error: #{error}")
            exit!
          end
        else
          result["data"]
        end
      else
        @logger.error("[GraphQL] Query failed: #{response.status} - #{response.body}")
        return nil
      end
    end

    def scene_slug(url)
      url = url.downcase.gsub(/\/$/, "")
      url.split("/").last
    end

    def parse_scene(response, url)
      return {} if response.nil? || response.empty?

      data = response["findOneVideo"]
      return {} unless data

      scene = {
        "title" => data["title"],
        "details" => data["description"],
        "studio" => {
          "name" => @name
        },
        "date" => "",
        "performers" => [],
        "tags" => []
      }
      date = scene["releaseDate"]
      scene["date"] = date.split("T")[0] if date
      if data["models"]
        data["models"].each { |model| scene["performers"] << model }
      end

      if data["tags"]
        data["tags"].each do |tag|
          next if @config.ignored_tags.include?(tag)
          scene["tags"] << { "name" => tag }
        end

        scene["tags"].push(*@config.fixed_tags) if @config.fixed_tags?
      end

      if data["images"] && data["images"]["poster"]
        max_width = 0
        data["images"]["poster"].each do |poster|
          scene["image"] = poster["src"] if poster["width"] > max_width
          max_width = poster["width"]
        end
      end

      scene["url"] = url if url

      scene
    end

    def parse_search(response)
      search_result = []
      return search_result if response.nil? || response.empty?

      data = response["searchVideos"]
      return search_result unless data

      data["edges"].each do |scene|
        next unless scene = scene["node"]
        next unless slug = scene["slug"]

        scene_meta = {
          "title" => scene["title"],
          "details" => scene["description"],
          "url" => "https://www.#{@id.downcase}.com/videos/#{slug}",
          "studio" => { "name": @name },
          "performers" => []
        }
        date = scene["releaseDate"]
        scene_meta['date'] = date.split("T")[0]

        if scene["modelsSlugged"]
          scene["modelsSlugged"].each do |model|
            scene_meta["performers"] << { "name" => model["name"] }
          end
        end

        if scene["images"] && scene["images"]["listing"]
          max_width = 0
          scene["images"]["listing"].each do |image|
            scene_meta["image"] = image["src"] if image["width"] > maxWidth
            max_width = image["width"]
          end
        end

        search_result << scene_meta
      end

      search_result
    end

    def get_video_query
      <<-'GRAPHQL'
        query getVideo($videoSlug: String, $site: Site) {
            findOneVideo(input: {slug: $videoSlug, site: $site}) {
                title
                description
                releaseDate
                runLength
                chapters {
                  video {
                    title
                    seconds
                  }
                }
                models {
                    name
                }
                images {
                    poster {
                        src
                        width
                    }
                }
                tags
            }
        }
      GRAPHQL
    end

    def get_search_query
      <<-'GRAPHQL'
        query getSearchResults($query: String!, $site: Site!, $first: Int) {
              searchVideos(input: { query: $query, site: $site, first: $first }) {
                  edges {
                      node {
                          description
                          title
                          slug
                          releaseDate
                          runLength
                        	chapters {
                            video {
                              title
                              seconds
                            }
                          }
                          modelsSlugged: models {
                              name
                              slugged: slug
                          }
                          images {
                              listing {
                                  src
                                  width
                              }
                          }
                      }
                  }
              }
          }
        GRAPHQL
    end
  end
end

def sort_by_length(studios)
  sorted = []
  studios.each do |studio|
    sorted << studio
  end

  sorted.sort_by { |studio| studio.length }.reverse!
end

studios = [
    VixenMediaGroup::Site.new("Blacked Raw"),
    VixenMediaGroup::Site.new("Blacked"),
    VixenMediaGroup::Site.new("Deeper"),
    VixenMediaGroup::Site.new("Slayed"),
    VixenMediaGroup::Site.new("Tushy"),
    VixenMediaGroup::Site.new("Tushy Raw"),
    VixenMediaGroup::Site.new("Vixen")
]

fragment = JSON.parse(STDIN.read)
search_query = fragment["name"]
url = fragment["url"]
logger = Stash::Logger.new

#sceneByURL
if url
  studios.each do |studio|
    if studio.valid_url?(url)
      scene = studio.get_scene(url)
      puts(JSON.dump(scene))
      exit
    end
  end
  logger.error("URL: #{url} is not supported")
  puts("{}")
  exit!
end

#sceneByName
if search_query && ARGV.include?("search")
  search_query = search_query.downcase
  list = []
  filter = []

  #  Only search on specific site if the studio name is in the search query
  # ('Ariana Vixen Cecilia' will search only on Vixen)

  # if the first character is $, filter will be ignored.
  if search_query[0] != "$"
    studios_sorted = sortByLength(studios)
    studios_sorted.each do |studio|
      if search_query.include?(studio.id.downcase)
        filter << studio.id.downcase
      end
      search_query = search_query.gsub!(studio.id.downcase, "")
    end
  else
    search_query = search_query.drop(1)
  end

  if filter
    logger.info("Filter: #{filter} applied")
  end

  logger.debug("Query: '#{search_query}'")

  studios.each do |studio|
    if filter
      unless filter.include?(studio.id.downcase)
        logger.debug("[Filter] #{studio.id} ignored")
      end
    end

    search = studio.get_search_result(search_query)
    list.merge!(search) if search
  log.debug(JSON.dump(list))
  puts(JSON.dump(list))
  exit
  end
end
```

halorrr commented:
@bnkai made a slight tweak to the logger class so that the methods are class methods instead of instance methods. Realized since I'm not initializing anything when we involve the logger it made more sense for them to be class methods like the configs.

I updated the example script I shared above to match that change.

Have you had a chance to test these out yet? I'm finishing up another scraper that will depend on these
